### PR TITLE
Made get_child support negative indexes

### LINKED ISSUE #1392: `Node.get_child(-1)` should return last child
**Describe the project you are working on:**
A game using GDScript.

**Describe the problem or limitation you are having in your project:**
This is a convenience issue. To get the last child of a node you need to do `get_child(get_child_count()-1)`, which is needlessly long.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
`get_child(-1)` should return the last child of a node. Could be theoretically expanded to negative indices like in Array, but from my experience the last child is much more often useful than second to last etc.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
> `get_child(-1)` should return the last child of a node.

The implementation is trivial, just check if the index is -1 and replace with child count -1 if yes.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
With this
> `get_child(get_child_count()-1)`

but it could be shorter.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It's about core method.

### PR DESCRIPTION
**Pull request by SekoiaTree**
PR for https://github.com/godotengine/godot-proposals/issues/1392. get_child was adapted to support negative indexes. Negative indices go between -1 to -(number of children). Not much to say.

*Bugsquad edit:* Closes godotengine/godot-proposals#1392.

### CODE REVIEW

**Comment by groud** [comment]
Could you update the documentation accordingly too ?

**Comment by SekoiaTree** [comment]
Oh yeah, sure. Will edit.

File: scene/main/node.cpp [hunk 1]
```diff
@@ -1327,8 +1327,10 @@ int Node::get_child_count() const {
 }
 
 Node *Node::get_child(int p_index) const {
-	ERR_FAIL_INDEX_V(p_index, data.children.size(), nullptr);
-
+	ERR_FAIL_COND_V(p_index >= data.children.size() || p_index < -data.children.size(), nullptr);
+	if (p_index < 0) {
```
Comment (Xrayez) [inline scene/main/node.cpp:1331]: I wonder whether this could benefit from branch prediction (#11461):

```suggestion
	if (unlikely(p_index < 0)) {
```

But I guess it wouldn't be too unlikely once people start to use negative indices. 😛

Comment (groud) [reply scene/main/node.cpp:1331]: Not really. Branch prediction has a significantly higher cost when the other case happen. So if there are cases where a negative index is used often, this could impact performances.

Comment (SekoiaTree) [reply scene/main/node.cpp:1331]: Even if people start using them, it's still rather unlikely, so I'll commit. Thanks!

Comment (groud) [reply scene/main/node.cpp]: No, honestly don't. 
Even if it's not used often, the code could be called several times in the frame. So it's important the performances stay correct in both cases.

**Review by groud — CHANGES_REQUESTED** [review]
Do not use unlikely here, if a negative index is called often in a frame it could hurt performances significantly.

**Comment by SekoiaTree** [comment]
Reverting.

**Comment by Xrayez** [comment]
Yeah I'm not really sure about `unlikely`, this also needs some benchmarks.

But then again, this feature is mostly requested as convenience, so it shouldn't impact performance in normal cases? I wouldn't worry about performance if you do go for convenience anyway. (?)

**Comment by SekoiaTree** [comment]
I'll try and do some benchmarks, but it's a single check, so it shouldn't impact much. One shouldn't use get_child enough to impact performance anyway.

**Comment by groud** [comment]
> But then again, this feature is mostly requested as convenience, so it shouldn't impact performance in normal cases? I wouldn't worry about performance if you do go for convenience anyway. (?)

In the general use case the impact shouldn't be that much. In the general use case, you don't call this often in a frame, so the performance impact should not be that important.

The problem is that if someone has a weird use case always going the unlikely way, the performances might be impacted significantly (not sure by how much, but from what I remember, going the unlikely way has a much significant impact than a normal `if` statement).

So basically, this is a trade of a little bit of performance in the general case for acceptable performances in specific use cases. Which is ok, as in the general use case, you don't need to call get_child() repeatedly.

**Comment by SekoiaTree** [comment]
Ran some tests, with the change, on a very simple scene (3 sprites) I get 10'000 get_child calls before a slowdown. Without the change... it's the same. The difference is practically unnoticeable.

**Comment by KoBeWi** [comment]
You could simplify the error check by making index positive before checking and then do `ERR_FAIL_INDEX_V` normally.

**Comment by SekoiaTree** [comment]
I thought of that; it doesn't work because the minimum index allowed is -(number of children) whereas the maximum index is (number of children - 1), meaning abs() wouldn't work.

**Comment by KoBeWi** [comment]
You don't need to use abs, it can be a condition like now.
```
if (p_index < 0) {
    p_index += data.children.size()
}
ERR_FAIL_INDEX_V(p_index, data.children.size(), nullptr);
```

**Comment by SekoiaTree** [comment]
Oh, I see. Yeah, I'll do that.

**Comment by SekoiaTree** [comment]
@groud could you remove/change your requested change? Since it's invalid now.

**Review by groud — APPROVED** [review]

**Comment by SekoiaTree** [comment]
Thanks! Just readded a line that I accidentally removed.

**Comment by akien-mga** [comment]
Thanks!

**Comment by SekoiaTree** [comment]
Thank *you* :D
